### DISCUSSION TITLE
New entity component: `event`

### DISCUSSION THREAD

**Discussion by frenck** [home-assistant/architecture#929] [Entity Models] [answered]
# Context

Events are signals that are emitted when something happens, for example, when a user presses a physical button like a doorbell. Events are fired and forgotten. They have no state. Therefore, if an event is missed, there is no way of telling the event has happened.

The use of events in Home Assistant is counter-intuitive. Users must know and learn multiple ways of workings events, all of which have pros and cons.

### Raw events

Working with raw events requires an end-user to puzzle/figure out events by listening to specific events using developer tools. With the data captured (the event data), one could make an automation using the event trigger and adding the acquired data as the "condition" for the trigger to fire.

The nice thing about working with raw events, is that it is very flexible. It allows for virtually any use case with a free-for-all event data set.

However, is not very user-friendly and requires more profound knowledge to work with Home Assistant. Also, there isn't a history available of when/which event happened. The sections below will show why we already didn't like this and have been "working around" this in several ways.

### Binary sensors

In some cases, we turn events into binary sensors. For example, creating a binary sensor for detected motion from a motion sensor. It turns the binary sensor "on" when motion is detected and "off" again after some timeout.

Although we do not allow for this pattern in general, we have some historical cases in our codebase, and above all, in some cases, the underlying libraries handle the timeout. It makes it look like it is a state, while it is not (ZHA does this, for example).

This pattern originates from making the events more accessible and easier to use in automations. The downside is that they are not a fit for binary sensors and are against our general design patterns.

However, it makes it easier for the user to use this in automations and provides a history of the state changes.

### Device automations

Device automations is a higher abstraction that was added a few years ago. It allowed device automation triggers and conditions on a device level. In many cases, this has been leveraged to handle events.

For example, you can detect a rotation of your Xiaomi Magic Cube using a device automation that is being fed/triggered by an event inside the core codebase.

This improved the user experience in automating devices providing events a lot!

However, it does require specific additional implementation from the integration. It also doesn't provide any history of the events that were fired. For example, you could not tell if the doorbell button was pressed in the last hour you've been away from home.

# Proposal

Short version: Introduce an `event` entity component.

### Goals

From the above context, the general solution we seek should fulfill the following goals:

- It should be easy to use by end-users directly from the UI. This is not limited to automations, but also scripts, dashboards, and other places.
- Should provide historical event information.

### Reasons to use an entity component

Comparing that to the context above, the binary sensor is the only one close to it. Therefore, the suggestion is to create an `event` entity to specifically handle these cases instead.

Some reasons why:

- Our users are familiar with the construct of entities.
- Entities provide history capabilities out of the box.
- Automation/script editors work with them and can be extended to provide suggestions/help with these entities.
- The entity component can provide device automation triggers, removing the need for integrators to provide them additionally.
- Device classes can be used to classify the type of device the entity provides, for example, a doorbell, motion, or button.
- Entities can be directly used in, for example, templates.
- The translation system can provide translation for entities into the users' locale.

### The event entity state

For the state of the event entity, we will be using the ISO 8601 timestamp of the last time the event fired, with microsecond precision and timezone information. For example: "2023-06-19T13:45:595681+00:00".

The timestamp of the state will not be settable by the integration. Instead, the integration can "fire" the event entity using a method, for which Home Assistant will handle setting the state.

The frontend will display/show this state, similar to a timestamp sensor entity.

### Event data & different types of events

Events often have additional information. For example, which button has been pressed, or the type of press (hold, double-press, single-press, release, ...), and sometimes even more.

To accommodate these better, an event entity has a list of possible events (`event_types`) it can provide (comparable to `options` in a `select` entity). It can provide the following:

- Translations for each of these events.
- Selections in the automation/script editor when addition triggers or conditions.
- More context in the event history for these events.

Events fires that are not known by the event entity will fail and cause an error logged to prevent unexpected events from being fired.

For additional data/information, the extra state attributes are available as per usual.

The frontend will display the type of event fired in the more information dialog of an entity. The extra state attributes are shown as usual.

### Restoring state

As the event entity has no state to retrieve, it will be restoring its previous state by default; without the need for an integrator to implement this specifically.

# Consequences

This will add more states to our state machine, having an effect on our recording into the database, but possibly also on our WebSocket.
Users will be able to use events using the concept of entities, which almost everybody is familiar with.

# Possible issues as a result of this proposal

- Breaking changes in entities. Some entities are expected to move from a binary sensor to an event entity to match the use case better. However, the binary sensor can be kept around as disabled by default to mitigate this.

- Some integrations may already have device automation triggers based on events. Adding an `event` entity in those cases, will cause duplicate "events" to select in device automation (the existing ones, and the ones provided by the existing implemented one). To possible approaches to resolving this:
  * Add a way for an integrator not to generate device automation for an `event` entity.
  * Add a way for an integrator to deprecate/hide the older event triggers from the user, while they remain working.

- We should rename the existing fire event service and event trigger in the UI to make clear those are for manually / advanced event handling.

# Future possible extensions

The following ideas have been left out of scope for this initial iteration. Mainly to keep the scope smaller and keep the architectural discussion more targeted at the core principle of this proposal.

- Helpers to detect single/double/triple presses, as inspired by ESPHome.

# Other considered alternatives

- Just use/allow binary sensors (extending a programmable/settable timeout?)
- Allowing main/sub-events is a common pattern in device automation triggers. A more straightforward approach is, IMHO, better (single level). 
- Use the existing sensor entity with, for example, an `event` device class.

## Proposal changes

- Bundling multiple (similar) inputs into a single entity using an `event_input` property. For example,  Hue Remote has 4 buttons, each would be an `event_input`. This idea was liked and disliked in the proposal discussion below.  
  The main issue is the UI / UX. It would require non-standard triggers/conditions to make this work.

**Comment by AngellusMortis** [comment]
Something to consider is that some events _do_ have state. Your motion detection is a great example. It varies based on manufacturer, but for UniFi Protect, motion and smart detection (motion that detects a specific object), but have a concrete start and end datetime. A motion event can last for multiple minutes if the camera keeps detecting it for that period of time. They also have additional metadata associated with them (event ID, confidence interval score, thumbnail, etc). 

So, the answer might be a "start event" and an "end event", but that may still just be easier to represent as a binary sensor.

**Reply by frenck** [reply]
> Something to consider is that some events do have state. 

In those cases, a binary sensor can be used. The `event` entity is meant to be used for resolving the case of not having an actual state.

**Reply by AngellusMortis** [reply]
I think it would then be important that we make similar event/binary_sensor entities appear consistent and the "same-ish" throughout HA then. If a camera uses a binary_sensor for motion, but a Zigbee sensor uses an event, a different UX may confuse the user as they may not understand the difference and why one is an event and the other is a binary_sensor.

**Reply by frenck** [reply]
We can't make it the same, as they aren't the same. That is the whole point of this proposal, see also the context set out above.

**Reply by AngellusMortis** [reply]
same-ish. And UX is different than internal representation. 

If the automation UI has "binary_sensor turned on" and "event fired" or it does not show history for events, it leaves us in the same place we are now and confuse users. A solution might be to use translations to handle this. i.e. if a binary_sensor has a device class of "motion", it says "binary_sensor started detecting motion" and "binary_sensor stopped detecting motion". Whereas a "motion" event would just have "event started detecting motion" (no "stopped" message/choice).

**Reply by frenck** [reply]
An event, in the Home Assistant context, is stateless (unlike a binary sensor). It isn't on or off. Thus, we cannot display it as on or off, we can only tell when it last fired.

**Reply by AngellusMortis** [reply]
In other words, an entity of the "motion" device class, should at as similar as possible in the UI to user regardless of if it is a binary_sensor or an event. Having them use different terminology or have a different UX would be confusing.

**Reply by AngellusMortis** [reply]
> Thus, we cannot display it as on or off, we can only tell when it last fired.

Okay, so you should have a last started instead (in the UI, it would still be fired internally) and the binary sensor can have a last started as well. Then you can also have a last ended for the sensor. You could also still show the history graph for both as well to show how often both things happen.

**Reply by frenck** [reply]
> Okay, so you should have a last started instead

There is no started. It is a single event fired. e.g., Someone pressed a button. Nobody starts pressing a button. It got an event the button was pressed.

**Reply by AngellusMortis** [reply]
@bdraco posted [his thought](https://github.com/home-assistant/architecture/discussions/929#discussioncomment-6219789), but we were talking and I think any device class we determine that _can have state_, like motion, should not ever be included as an event. That will solve this whole problem and discussion here.

This can bad tracked in a different proposal (only important part is that we are saying motion is _not_ an event entity), but the idea both of us liked was binary_sensors with a "feature" that allows for an optional end time. If it does not support an end time, it uses an integration provided timeout to automatically turn off. 

i.e. if an API implements start _and_ end events for motion, it uses a binary_sensor exactly as it is now. If an API _only_ implements start events for motion, it uses a binary_sensor that automatically turns off after a set amount of time (say 30 seconds).

We cannot control how APIs are implemented by third parties, but having two different types of "motion detection" (or similar) is not a good solution. It creates the UX problem I have been trying to mention in this thread.

**Reply by allenporter** [reply]
Events with a duration may be like calendar events for what it's worth, if it helps to have a broader set of use cases.  Right now the calendar entity does double duty as an agenda providing a schedule and a binary sensor like entity and I've been considering a helper to tease apart.  (One difference is users want offsets, but that detail may be encapsulated in an entity)

(Not that this thread is the right place to figure that out. Something to revisit in the future perhaps)

**Reply by frenck** [reply]
I'm not following any of the above. This is not about those cases. This is about bringing Home Assistant events into the UI as an event. You seem to be stuck on a specific use case for a specific motion thing, which is not what this architectural discussion/proposal is about.

IMHO, causing quite a bit of unrelated noise.

**Reply by allenporter** [reply]
> I'm not following any of the above. This is not about those cases. This is about bringing Home Assistant events into the UI as an event.

Yes, I think it's pretty clear what is in scope for this proposal and what is not, as this is explicitly stated many times here.

It is healthy to discuss topics at the edge of the proposal as it may spawn off future proposals, or help understand this proposal's scope better by what's not in it.  IMO the sign of a good proposal is that it gets people thinking and discussing.

**Reply by frenck** [reply]
> It is healthy to discuss topics at the edge of the proposal as it may spawn off future proposals

I agree, but it can also can cause focus to get lost and derailment of what is actually the core of the conversation. Considering the quite large wall of discussion on something out of scope in this proposal, I really think derailment is happening and want to stop that.

../Frenck

**Reply by AngellusMortis** [reply]
The whole entire point of this thread has been the user experience impact of having multiple ways to display the same type of thing to the user. I do not like how you are dismissing it just because you apparently do not think user experience is important.

If it is decided as part of this discussion to make "motion" a device class of an event, that will set precedent for not it being separate from "motion" binary sensors. If that is done, I fear we are putting the burden of understanding on the user for the fact that different APIs implement the same concept (motion detection) differently. If user has a Nest Camera and an Aqara PIR sensor and motion is presented as an event by the Nest Camera and a binary sensor by the Aqara sensor, the is confusing. Motion is motion and it is on HA to present it in a unified manner.

**Reply by frenck** [reply]
> I do not like how you are dismissing it just because you apparently do not think user experience is important.

I stop any form of further interaction with you at this point.

**Comment by parautenbach** [comment]
To support this proposal, I'd like to link https://github.com/home-assistant/core/issues/58056 here as a real example of what Frenck mentioned. I'm keen to see how this proposal develops and what an improved implementation would look like.

**Comment by Jc2k** [comment]
I love this.

For homekit_controller, I would love to drop our custom device automation triggers entirely and replace them with events - it would simplify our code quite a bit and I think it would actually remove some races. Maybe for that scenario `async_get_triggers` can return `[]` and `async_attach_trigger` can some how cause a migration from a custom device automation trigger to the equivalent trigger for the new `event` entity. And then we could follow the cadence of ADR-0021, and keep those migrations in place for at least 6 months?

**Reply by frenck** [reply]
I'm not sure how to handle migrations in this regard, as also listed in the "Possible issues as a result of this proposal" section of the proposal. The thing is, we can't update the automations people have, yet we don't want to break them either.

I've listed two possible solutions I can think of, but I think we need to keep the "old" ones around forever 😟 (but maybe hide them from the UI to prevent duplicates/new usages).

**Comment by bdraco** [comment]
I think this proposal makes sense for doorbells and momentary events.  ✅ 

The motion sensor case is more complex, and I think its best to address that in a separate proposal as I have at least the following types of motion sensors:

- Motion sensor that fires event when it first detects movement and fires an event when movement stops
- Motion sensor that continuously fires events during movement
- Motion sensor that records the start and end of movement

> In some cases, we turn events into binary sensors. For example, creating a binary sensor for detected motion from a motion sensor. It turns the binary sensor "on" when motion is detected and "off" again after some timeout.

I don't think this is a good fit for motion sensors that have a duration. I'm not sure if its intended to be from reading the proposal so it may just be the example above muddying the water a bit here.

**Reply by pvizeli** [reply]
Agree. I think button/remote are not entirely fit, we should make a button/remote component for that

**Reply by pvizeli** [reply]
hmm, I guess it could work if the remote/button is an device and have this event entities for long/short press ect.

**Reply by frenck** [reply]
> I don't think this is a good fit for motion sensors that have a duration.

Correct, but during that known duration, we know the state. In these cases, a binary sensor can be used (we've allowed for that in the past as well). However, the magic word in this part is: known duration.

**Comment by allenporter** [comment]
I am supportive. This proposal works for all nest camera events which are currently implemented as device triggers.  The special parts are the event type (e.g. person, motion, camera, doorbell chime, etc) and an event id which can be an extra state attribute.

On the event data: I suspect some events may have data that is useful to expose in an automation but is not useful to write to the recorder for the long term. May be worth considering if thats something to omit or support and if the existing recorder exclusion can handle that.

**Reply by frenck** [reply]
Integrations can exclude attributes from being recorded by implementing the recorder platform.

**Reply by Rudd-O** [reply]
This would indeed be awesome.  Device triggers often make users (made me!) create helpers whose state is then altered by automations.  That feels backwards!  It feels to me like automations should be driven by entity changes, rather than needing automations to alter the states of entities.

**Comment by shbatm** [comment]
I really like this idea and can think of several use cases where I already have work arounds in my automations for it. ISY, Insteon, Zigbee, and Z-Wave JS would all have use cases. 

One thought: is calling it `event` going to introduce confusion with the underlying/existing events on the event bus? Or is the intent for every current underlying event on the bus becoming an entity? 

Just wondering if it should have a different name like `action` or `instant` to prevent confusion (`trigger` would also introduce conflict).

**Reply by frenck** [reply]
`action` is used by automations and scripts and have a different meaning.

**Reply by shbatm** [reply]
Using `control` was another thought, but yeah, it's tough finding something that is not conflicting or confusing.

**Comment by frenck** [comment]
So, this architectural proposal came up in our weekly meeting, and we stuck on it waaay to long 😬

On case to consider, that needs a good design is, for example, a Hue remote. It has 4 buttons a user can press (on, off, brightness up and down), and each button has multiple event types (for example, single and double click).

The discussion ended up into:

- Should this remote be a single `event` entity? Or multiple entities (one for each button)?

In case of a single entity, in the current proposal, we might end up with an event list like: `on_single`, `on_double`, `up_single`, `up_double`, `down_single`, `down_double`, `off_single`, and `off_double`.

This will grow hard if there are more types of events and buttons to press (looking at you aqara cube, but let's not design it around those things). Which can lead to huge selections lists in the UI.

So, one of the possible suggestions is to bring in `inputs` besides `events`.

In this case, `inputs` are the buttons of the Hue remote. In that case, an event entity will be fires based on the input used, and the event fired.

This design does assume that each `input` has the same list of `events`.

In case no inputs are defined (None), it is assumed to be a single/main input. Like a doorbell with a single button.

**Reply by shbatm** [reply]
My vote would be multiple `event` entities for devices with distinct different buttons. The `input` suggestion feels like mixing `event` entities and the list of possible event states. Also `input` seems like it would cause confusion with the `input_*` helper entities. 

In addition to your Hue remote example, another would be for Z-Wave scene controllers (`event.myswitch_scene_controller_1, _2, _3`) or other protocol remotes.

**Reply by Jc2k** [reply]
I think an entity per event source (so 4 buttons on a hue remote = 4 entities) could work well from a UX pov.

**Reply by shbatm** [reply]
Thinking more on it, should we also leave it up to the specific integration on how to break them up... For example, using the hue remote does On/Off and Brightness Up/Down just become 2 entities, instead of 4?

**Reply by frenck** [reply]
@shbatm Well, we wanted to set guidelines to prevent wild implementations/things from being created :)

**Reply by shbatm** [reply]
Understood. So should the standardization be something along the lines of: (a) "Each distinct instantaneous input on a device should be represented as separate `event` entities." (Hue example = 4 entities) or (b) "Each input functional group should be represented as separate `event` entities (Hue becomes power and brightness / 2).

**Reply by Jc2k** [reply]
Personally I prefer (a).

(B) kinda works for vendor specific integrations because there probably aren't many variations to deal with. It's not onerous to define those groups in the integration code. But for integrations that are multi vendor it gets kinda onerous. 

E.g. the hue remote as an example. If you use it with homekit_controller you get a group of four buttons. And you know what order they are in. If hue made another remote with a different button configuration the current code would still work on release day. To do (b), we'd have to maintain a quirks DB and devices wouldn't meet (b) until they were added to it. But for (a) they would work on release day. And then if we later added a device to such a grouping quirk DB, would it be a breaking change for anyone that already had it paired?

I imagine zha, zwave, z2m and eventually matter would be in a similar position?

**Reply by shbatm** [reply]
Sounds good to me. I'd imagine it should be what the device 'naturally' presents as discrete inputs so there isn't a need for maintaining a separate quirks list. 

Tweaking my suggestion for 'a' to: 

>"Each distinct instantaneous input endpoint on a device should be represented as separate event entities."

So if you get 4 distinct controls/endpoints presented from the device, you have 4 entities. I have some zigbee remotes that don't separate the clusters and just send on/off/dim up/dim down/dim stop as 1 'action' output, so I'd only expect a single 'event' entity for that one. This would also work for the likes of Insteon which just presents a single 'command' event for each device.

**Reply by shbatm** [reply]
> In case of a single entity, in the current proposal, we might end up with an event list like: on_single, on_double, up_single, up_double, down_single, down_double, off_single, and off_double.

Just as an FYI, I wanted to double check on Z2M (my biggest use case w 100+ devices) and that is currently how it's structured for the action entity: https://www.zigbee2mqtt.io/devices/E1524_E1810.html#action-enum

**Reply by farmio** [reply]
For a single push button, would it be possible to have `press` and `release` actions in one event entity or would that be 2 entities? 
An example would be `dim-up-start` and `dim-up-stop` for a button that is expected to dim up as long as it is pressed. Or should such thing be a binary_sensor?

**Reply by ScottG489** [reply]
On the topic of buttons, would we consider these events to be the read-only equivalent of a `button` entity? Similar to how we have `switch` and `binary_sensors`. It seems like there's overlap in the core purpose, and the current behavior of buttons is also to use a timestamp as it's state.

I say this only to point out that having an example already implemented in the codebase could help with imagining how we might want this to look and potentially borrow from.

Though, thinking about it a bit more, I don't know how much overlap even a `switch` or `binary_sensor` really have. So a comparison may not be appropriate. Still, I believe the `button` is the only existing entity which also would represent an occurrence.

**Reply by jlpouffier** [reply]
The fact that each `input` on the `inputs` list needs to support the same `events` may be too restrictive.

A 4-button remote will have the same `events` for each button... So in this case, yes having an `input` is definitely a good solution to discriminate buttons and to make the `events` list shorter 

But if the integration of your 4-button remotes wants to support different types of events such as `low_battery` or `weak_connection` (I am making these events up just to make a point), then the coupling `inputs x events` will force the integration to provide a different `event` entity.

For the Aqara Cube, I don't see what could be the possible value of `inputs`, and I think the integration will have to provide different `event` entities.

In an effort to provide guidance and consistency across integrations, I would keep it simple and drop the `inputs` concept and guide the integration to provide multiple `event` entities

JLo

**Reply by frenck** [reply]
> On the topic of buttons, would we consider these events to be the read-only equivalent of a `button` entity?

@ScottG489 That is what this proposal is about, please read the OP.

**Reply by frenck** [reply]
> For the Aqara Cube, I don't see what could be the possible value of inputs

Correct, it will be a single input... the cube.

However, a remote like Philips Hue, it would be the different buttons.
The reasons for this approach could be: Let's say we have a remote with 32 buttons, it would cause 32 event entities; using inputs it could be a single event entity.

**Reply by ScottG489** [reply]
> @ScottG489 That is what this proposal is about, please read the OP.

@frenck I took the mentions of buttons to mean physical buttons, not the entity HA. I was referring to using implementation details from the `button` entity to help model what we want this entity to look like and how to behave.

You can't always map a physical button to a `button` entity because you may not be able to reproduce the behavior of the physical button from HA. So that's an instance of where this `event` entity would fill a need well.

**Reply by jlpouffier** [reply]
> The reasons for this approach could be: Let's say we have a remote with 32 buttons, it would cause 32 event entities; using inputs it could be a single event entity

But from what I understand, and please correct me if misunderstood something, if the 32 button remote also expose a `low_battery` event for example, then **at least** this event would need to be on a different entity, because it's entity_type will lt be the same then the 32 others 

Correct?

It will be 
- one event entity with 32 inputs and two event_type (let's say short_press and long_press)
- one other entity event for the low_battery event

**Reply by frenck** [reply]
> It will be

Correct :)

**Reply by jlpouffier** [reply]
Cool ! Glad I understand 

Then this means that the "input solution" is not a perfect solution for the problem "One physical device = One event entity"

I think in the end it's fine because it still helps reduce the number of entities to a minimum.

**Reply by Rudd-O** [reply]
If `button` entities aren't adequate for the type of remote (and there are reasons why that might be), then we'll definitely need an event input axis (up, down, left, right, center...) and an event kind axis (press, release, perhaps double-click or long-press although these can be built or adapted at a higher level) as well.  The event itself could have an attribute such as "event kind is release, and press duration was 0.45s" to aid in the construction of these higher-level meanings.

**Reply by jlpouffier** [reply]
I just had a shower though and want to share it here ( I'm am not aware of any remote supporting this)

Pressing on two buttons at once may be simpler to represent with two independent events than with one event containing two "kind/event_type"

When we think about it, the different buttons are fully independent...

**Reply by marcelveldt** [reply]
I'm not so sure if I agree because in most cases that is a different action. So it completely depends on the device and what it supports. My personal feeling is that one physical device should be one "event" entity, even if it has multiple actions like multiple knobs or even a rotary.

But on the other hand there will also be cases where each individual button on a device has its own dedicated task and then I'm not so sure anymore. Maybe leave that part on the implementation/integration

**Reply by frenck** [reply]
I've updated the proposal and removed the `event_input` attribute as discussed in this part of the discussion. It makes the entity harder to use and the UX harder to achieve, besides, in the above there seems to be no consensus either.

**Comment by jlpouffier** [comment]
One very small feedback from my end. 
Having an entity of type `event` that has, as an attribute, a list of possible `events` it can provide feels weird and counterintuitive.
I would propose to rename the list.
No strong opinion on the renaming, but I would propose `types` or `event_types`


> _An event got fired on the entity `event.my_remote`, its type is `long_press`_ 

Seems more intuitive


JLo

**Reply by frenck** [reply]
`type` is a reserved keyword in Python, hence I avoid those terms in general.

**Reply by jlpouffier** [reply]
Oh no problem! 
As stated, I have no opinion on the renaming.
What bothers me is that we are thinking about creating `an event that has a list of events`. 
That's the point of my comment 

I can propose a few other names
- category
- class


> _An event got fired on the entity `event.my_remote`, its category is `long_press`_

Still intuitive to me

**Reply by Jc2k** [reply]
My mental model for this is that it's a sensor/binary_sensor but stateless. So it's more like stateless_sensor.

**Reply by frenck** [reply]
We use `event_type` for the event trigger: <https://www.home-assistant.io/docs/automation/trigger/#event-trigger>

So that might be a good match, as suggested by @jlpouffier

**Reply by Rudd-O** [reply]
s/type/kind/

**Comment by marcelveldt** [comment]
So, I'm 100% positive about a way to "visualize" device events to users as they are now lost and actually create issues reports that their device is missing entities. But it is not as simple. I've been working on device events in Z-Wave (JS), Hue and Matter and here are my 2 cents:

1) I think we make it way too restrictive if we require the list of all possible events to be provided upfront (and error on other errors). That is bad because I've seen way too many examples where new events were added to an existing device afterwards. It would be very nice to match some default events maybe but we should not require it. 

2) In my opinion it should be up to the integration to determine if multiple entities are needed for a device but in 99% if the cases it will be a single device emitting events and thus it will be a single entity. In most cases you cant have different events at the same time (but soon next to eachother). 

Imo we should have an event entity with its state set to a timestamp of the last event and a state attribute(s) with details about the event, mandatory would be 'event' with some sort of integration specific event (some string) and optional attributes only if the integration can provide them, such as 'event name' or 'event duration' etc. etc. Another optional state attribute would be 'events' which would hold a list of all possible events but only if the integration is capable of providing those.

I was just about to add device events support for the Matter integration after fixing support for it in the server.
How about we use Matter as a test case to test out this new entity ? Before we go wild and introduce it for all existing intregrations and even start removing device triggers. Or just add it to a few integrations (and keep the existing events and device triggers) so users can give some feedback how this works for them.

**Reply by frenck** [reply]
> I think we make it way too restrictive if we require the list of all possible events to be provided upfront (and error on other errors). 

I strongly disagree on that. We should not handle things from upstream that we are not familiar with. We have been implementing these patterns all over the place the last years, for example, enums on sensor and the options in select entities. You are proposing the anti-pattern now, I think we could allow the suggested, only if we remove it on other places too (which I would still not agree on).

Without it, we also won't be able to tell the user what the options are, what they can expect and we can provide translations either.

> That is bad because I've seen way too many examples where new events were added to an existing device afterwards. It would be very nice to match some default events maybe but we should not require it.

In that case, we need to add support. This is not bad, IMHO it is the opposite, we can add actual support in a expected way.

> In my opinion it should be up to the integration to determine if multiple entities are needed for a device but in 99% if the cases it will be a single device emitting events and thus it will be a single entity. 

That is what my here is about: <https://github.com/home-assistant/architecture/discussions/929#discussioncomment-6221849>

We actually want to set guidelines for this, to prevent this becoming a zoo of variations harming the user experience. Hence I do not agree on that comment.

> mo we should have an event entity with its state set to a timestamp of the last event and a state attribute(s) with details about the event, mandatory would be 'event' with some sort of integration specific event (some string) and optional attributes only if the integration can provide them, such as 'event name' or 'event duration' etc. etc. Another optional state attribute would be 'events' which would hold a list of all possible events but only if the integration is capable of providing those.

How is this different from the proposal?

../Frenck

**Reply by marcelveldt** [reply]
OK, relax :-) I think you've misread my proposal (or I've used the wrong words then). I agree fully with providing the possible events as a list to make it absolute clear to the user what events can be expected (and to create a nice automation GUI for it).

I just have 2 concerns:

- I don't think that requiring a separate entity for **each button** is a good idea. You could end up with tons of buttons. Maybe that is dependent of the usecase/integration though. An example is the Hue Tap Dial switch which has 4 buttons and a rotary encoder. You could decide to split that out into 5 event entities per device or just one per device.

- How do we catch the cases where the event itself is dynamic ? Keep that as an integration specific event submitted on the eventbus (as the usecase is perhaps considered advanced then anyways). An example would be some text inputted on some sort of device etc. Not sure about exact usecases right away but I'm sure they will present themselves at some point. Better to be ahead and describe what is NOT supposed to be supported by this entity ;-)

So maybe its as simple as "only use the event entity if the event is predictable and we know all values upfront".

**Reply by frenck** [reply]
> OK, relax :-) 

You should see me right now :) I'm enjoying an ice-tea and working in my swimming pants as I just took a dive :D I just miss the palm tree 🏖️ 😄 

> I don't think that requiring a separate entity for each button is a good idea. You could end up with tons of buttons. Maybe that is dependent of the usecase/integration though.

👍 I think that is the general consensus/signal here too: <https://github.com/home-assistant/architecture/discussions/929#discussioncomment-6221849>; however, we do want to set some guidelines for it. For example, it would not improve the user experience if one integration provides a 4 button remote as a single event entity, while the hue integration would provide 4 entities for each button of the Hue remote :)

> How do we catch the cases where the event itself is dynamic ?

I don't think this entity would be suitable for such situations.

> An example would be some text inputted on some sort of device etc.

That would be a sensor, not an event.

That said, I get your point, but it is easier to loosen it later than making it more strict later. Thus a strict start would IMHO be a better approach. If a use case appears, we can always discuss and look into the need for a non-predefined set of event types.

**Reply by Rudd-O** [reply]
I don't think there is a problem with having tons of event entities.  A PC keyboard device should have 105 event entities.  My own Bluetooth LE keyboard based on an ESP device does indeed have a button for each "key" you can depress (these are obviously exposed to HA using the button entity since that's the only thing that currently exists).  Users can always disable or enable these entities — perhaps the UI may need some souping up to ease batch operation of this particular change of config, but that's all.

**Reply by marcelveldt** [reply]
I do think that is a problem, users will get frustrated by it. But... maybe we're confusing UI and entitites. What if it are indeed multiple entities but we can logically group them in the UI so it makes sense again to the user ?

**Comment by Rudd-O** [comment]
This is exactly how I always envisioned events would work.  The current event system, which requires me to reverse engineer their format based on a stream of WebSocket events, is a nightmare.  I could cut half my use of Node-RED if the objects I have in my home gained event entities.

Will event states also have formally-defined attributes that can be added by the integrator?  I'm thinking the rotary encoder rotation of my Philips Hue Tap dial emits an angle of rotation, which is important information.  Currently I use this to decide (via a very complex Node-RED flow) by how much to raise the volume of my home theater, or how far to seek in the current YT video.

EDIT: in the case of "fake" binary sensors (which really are events) becoming deprecated, a repair can be used to indicate that there exist automations or scripts which are using these binary sensors.  Furthermore, a new type of binary sensor "helper" could be created, which would aid users in creating binary sensors out of events, complete with customizable timeout.

**Comment by Rudd-O** [comment]
>I'm not so sure if I agree because in most cases that is a different action. So it completely depends on the device and what it supports. My personal feeling is that one physical device should be one "event" entity, even if it has multiple actions like multiple knobs or even a rotary.

Hm.

Why tho?

Take the case of the Tap dial.

The device has four buttons and a rotary encoder.

Logically, it would follow that the device should present four button entities and an event entity that produces states for the rotary encoder.  Or maybe four event entities one per button and an event entity for the rotary encoder.

I don't see why we need to shoehorn all of a device's gizmos into a single event entity.  What benefit does that provide?  That is what we already do today — shoehorn all events through a single pipe.  This is part of what makes automations hard to build — you end up having to write complicated logic to "steer the event".  Much easier if every gizmo is its own event source entity.

Of course, it will ultimately be the decision of the implementer, but Home Assistant should not recommend to implementors that devices shoehorn every one of its event-producing gizmos into a single entity.

**Reply by frenck** [reply]
Can you please use threads here? As in, it is unclear where this is a response too...

**Reply by Rudd-O** [reply]
Yes, understood.  Please see below for why my reply didn't thread.

**Comment by Rudd-O** [comment]
Yes, we sort of already do this for some things (config and diagnostic entities get this treatment on the device page).

**Reply by marcelveldt** [reply]
Yeah, I think that is going to be the answer. Be able to group/represent them again as something that makes sense to the user. 
It would even be super cool if you could show some picture/graphic of the device and know what button is being pressed

**Reply by frenck** [reply]
Can you please use threads here? As in, it is unclear where this is a response too...

**Reply by Rudd-O** [reply]
Sorry, but I replied over e-mail as I was on my way to work, and Github isn't smart enough to thread.

(I don't have access to Github on my phone.)

**Comment by frenck** [comment]
I've updated the architectural proposal and think I've now processed most/all comments. 

Please let me know if there is something missed / unclear or things that should be improved. Meanwhile, I'll move forward with starting an implementation.

../Frenck

**Reply by Jc2k** [reply]
Sorry if I missed it while re-reading (I'm walking to office and on phone). What's the final consensus on the hue remote case? I saw it's not "event_input".

Eg is it permitted to have "event.button1", "event.button2", "event.button3", "event.button4" all attached to one device? And then single/double etc would be extra data on those events?

I think that would technically work with the existing UI infrastructure?

**Reply by marcelveldt** [reply]
Yes, each physical button will then have a separate entity. In the future the UI should maybe present these entities into one grouped entity control or something.

**Comment by frenck** [comment, marked as answer]
Draft implementation: 

- Link to architectural proposal: https://github.com/home-assistant/architecture/discussions/929
- Link to code base pull request: https://github.com/home-assistant/core/pull/96797
- Link to frontend pull request: https://github.com/home-assistant/frontend/pull/17332
- Link to brands pull request: https://github.com/home-assistant/brands/pull/4555
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/28263
- Link to developer documentation pull request: https://github.com/home-assistant/developers.home-assistant/pull/1849

![CleanShot 2023-07-20 at 17 52 32](https://github.com/home-assistant/frontend/assets/195327/9b4d0667-7620-4ccc-95af-dce4e1c02be3)

**Reply by frenck** [reply]
All PRs are ready for review

**Reply by frenck** [reply]
Implementation has been finished and merged for 2023.8.
